includes/Classifai/Providers/OpenAI/ChatGPT.php

```diff
@@ -0,0 +1,510 @@
+
+	/**
+	 * OpenAI ChatGPT constructor.
+	 *
+	 * @param string $service The service this class belongs to.
+	 */
+	public function __construct( $service ) {
+		parent::__construct(
+			'OpenAI',
+			'ChatGPT',
+			'openai_chatgpt',
+			$service
+		);
+	}
+
+	/**
+	 * Can the functionality be initialized?
+	 *
+	 * @return bool
+	 */
+	public function can_register() {
+		$settings = $this->get_settings();
+
+		if ( empty( $settings ) || ( isset( $settings['authenticated'] ) && false === $settings['authenticated'] ) ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Register what we need for the plugin.
+	 *
+	 * This only fires if can_register returns true.
+	 */
+	public function register() {
+		add_action( 'enqueue_block_editor_assets', [ $this, 'enqueue_editor_assets' ] );
+	}
+
+	/**
+	 * Enqueue the editor scripts.
+	 */
+	public function enqueue_editor_assets() {
+		$settings = $this->get_settings();
+
+		// Don't load our custom post excerpt if excerpt functionality isn't turned on.
+		if ( ! isset( $settings['enable_excerpt'] ) || 1 !== (int) $settings['enable_excerpt'] ) {
+			return;
+		}
+
+		// This script removes the core excerpt panel and replaces it with our own.
+		wp_enqueue_script(
+			'classifai-post-excerpt',
+			CLASSIFAI_PLUGIN_URL . 'dist/post-excerpt.js',
+			get_asset_info( 'post-excerpt', 'dependencies' ),
+			get_asset_info( 'post-excerpt', 'version' ),
+			true
+		);
+	}
+
+	/**
+	 * Setup fields
+	 */
+	public function setup_fields_sections() {
+		$default_settings = $this->get_default_settings();
+
+		// Add the settings section.
+		add_settings_section(
+			$this->get_option_name(),
+			$this->provider_service_name,
+			function() {
+				printf(
+					wp_kses(
+						/* translators: %1$s is replaced with the OpenAI sign up URL */
+						__( 'Don\'t have an OpenAI account yet? <a title="Sign up for an OpenAI account" href="%1$s">Sign up for one</a> in order to get your API key.', 'classifai' ),
+						[
+							'a' => [
+								'href'  => [],
+								'title' => [],
+							],
+						]
+					),
+					esc_url( 'https://platform.openai.com/signup' )
+				);
+			},
+			$this->get_option_name()
+		);
+
+		// Add all our settings.
+		add_settings_field(
+			'api-key',
+			esc_html__( 'API Key', 'classifai' ),
+			[ $this, 'render_input' ],
+			$this->get_option_name(),
+			$this->get_option_name(),
+			[
+				'label_for'     => 'api_key',
+				'input_type'    => 'password',
+				'default_value' => $default_settings['api_key'],
+			]
+		);
+
+		add_settings_field(
+			'enable-excerpt',
+			esc_html__( 'Generate excerpt', 'classifai' ),
+			[ $this, 'render_input' ],
+			$this->get_option_name(),
+			$this->get_option_name(),
+			[
+				'label_for'     => 'enable_excerpt',
+				'input_type'    => 'checkbox',
+				'default_value' => $default_settings['enable_excerpt'],
+				'description'   => __( 'A button will be added to the excerpt panel that can be used to generate an excerpt.', 'classifai' ),
```

Comment:
Even more automated would be to generate that excerpt on publish (and perhaps on update) versus requiring someone clicks a button.  On one hand we automate things completely while on the other you limit the number of calls to OpenAI.  I'm curious for thoughts from folks on best initial approach to take here?

Reply:
After reading your thoughts in the PR description @dkotter I'm thinking an alternate might be generating the excerpt automatically on publish (or perhaps doing that heading into the pre-publish panel and displaying the info there and allowing someone to change the excerpt or re-generate a new excerpt from OpenAI.  Once the post is published, having the "Generate Excerpt" or "Re-generate Excerpt" (if an excerpt already exists) could allow for someone to tweak the excerpt if they're not liking what's present.

Reply:
Probably worth talking this through to align on best workflow for this. I will say I don't think we want to have it automatically generated on publish, as that doesn't give you a chance to review the excerpt before that goes live. I like the idea of utilizing the pre-publish panel but I think we would still need to have a button that is clicked to send off the API request, so workflow won't be much different

Reply:
Hi @jeffpaul @dkotter, In response to this possibilities, I've created a PR: #408 off of this one that would add the excerpt in the Pre Publish Panel to support regeneration before publishing. I also added some minor improvements.